### PR TITLE
feat: ParamType from TypeApplication/TypeDeclaration in order to generate fn selectors without code generation

### PR DESCRIPTION
**Pull request by segfault-magnet**
closes: #617 
Assigning @digorithm since I'm not available tomorrow. Will reassign if this isn't merged the day after.

Good news and bad news.
Bad news first, we have to resolve generics again. 

Good news I've done the work already in a previous - now deleted - generics iteration.

You'll notice similarities between this and the `resolved_type`. I have a good feeling we're going to be able to unify these two into a single abstraction in the near future.

Currently `resolve_type` is only interested in the type. E.g. for some struct or enum it doesn't care about their fields/variants only how their type signature might look like. 

Whilst ParamTypes, and fn selectors by extension, care about everything. Therein some low-hanging fruit might be found.

### CODE REVIEW

**Comment by ra0x3** [comment]
@segfault-magnet 
- Specifically with regard to [this comment](https://fuellabs.slack.com/archives/C02CX4JVCSJ/p1665238091024929?thread_ts=1665187178.468689&cid=C02CX4JVCSJ), how would I use this new work? 
- I'm looking over the PR, but might be quicker to just ask you

File: packages/fuels-core/src/code_gen/function_selector.rs [hunk 1]
```diff
@@ -2,8 +2,8 @@ use crate::utils::first_four_bytes_of_sha256_hash;
 use crate::ByteArray;
 use fuels_types::param_types::ParamType;
 
-/// Given a `ABIFunction` will return a ByteArray representing the function
-/// selector as specified in the Fuel specs.
+/// Given a function name and its inputs  will return a ByteArray representing
+/// the function selector as specified in the Fuel specs.
 
 pub fn resolve_fn_selector(name: &str, inputs: &[ParamType]) -> ByteArray {
```
Comment (digorithm) [inline packages/fuels-core/src/code_gen/function_selector.rs:8]:
```suggestion
/// Given a function name and its inputs  will return a ByteArray representing
/// the function selector as specified in the Fuel specs.
pub fn resolve_fn_selector(name: &str, inputs: &[ParamType]) -> ByteArray {
```

Shouldn't this comment be bound to `resolve_fn_selector`? Seems like there was a line between them.

File: packages/fuels-types/src/param_types.rs [hunk 2]
```diff
@@ -88,110 +96,273 @@ impl ParamType {
...
+    fn determine_generics_for_type(
+        type_application: &TypeApplication,
+        types: &HashMap<usize, TypeDeclaration>,
+        parent_generic_params: &[(usize, Type)],
+    ) -> Vec<(usize, Self)> {
+        let type_decl = types.get(&type_application.type_id).unwrap();
+        match &type_decl.type_parameters {
+            // The presence of type_parameters indicates that the current type
+            // (a struct or an enum) defines some generic parameters (i.e.
+            // SomeStruct<T, K>
```
Comment (digorithm) [inline packages/fuels-types/src/param_types.rs:195]: Missing matching parenthesis:

```suggestion
            // The presence of type_parameters indicates that the current type
            // (a struct or an enum) defines some generic parameters 
            // (i.e. SomeStruct<T, K>).
```

File: packages/fuels-types/src/param_types.rs [hunk 3]
```diff
@@ -278,4 +451,698 @@ mod tests {
...
+        assert_eq!(
+            result,
+            ParamType::Tuple(vec![ParamType::U8, ParamType::String(15)])
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn ultimate_example() -> Result<(), Error> {
```
Comment (digorithm) [inline packages/fuels-types/src/param_types.rs:794]: Love the test name 😂

**Review by digorithm — CHANGES_REQUESTED** [review]
This looks awesome, comments and tests are super on point! Left some nits/suggestions, plus one more thing: Can you write a very short section about this new feature in the book? This is very useful for SDK users diving into deeper levels (as opposed to writing tests, deploying, etc.), and it would be a PITA to learn about this functionality without an easy to find section in the book 😅

**Comment by ra0x3** [comment]
> > @segfault-magnet
> > 
> > * Specifically with regard to [this comment](https://fuellabs.slack.com/archives/C02CX4JVCSJ/p1665238091024929?thread_ts=1665187178.468689&cid=C02CX4JVCSJ), how would I use this new work?
> > * I'm looking over the PR, but might be quicker to just ask you
> 
> Not sure if this is what you're asking for (can't find the comment you linked), but assuming that "how would I use this new work" you mean how to use the new stuff introduced in this PR: Once your parse the JSON ABI, you're gonna have a bunch of `TypeApplication`s. Then you just have to create `ParamType`s for each of them with `let result = ParamType::try_from_type_application(&type_application, &type_lookup)?;` where `type_lookup` is the map of all types (you also have those once you parse the JSON ABI).

That answers my question perfectly, thanks @digorithm

File: docs/src/debugging/function-selector.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Function selector
```
Comment (mohammadfawaz) [inline docs/src/debugging/function-selector.md:1]: This file can use some Markdown linting.

File: examples/debugging/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        let inputs = a_fun
+            .inputs
+            .into_iter()
+            .map(|type_appl| ParamType::try_from_type_application(&type_appl, &type_lookup))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let selector = resolve_fn_selector(&a_fun.name, &inputs);
+
+        assert_eq!(selector, [0, 0, 0, 0, 39, 152, 108, 146,]);
+
```
Comment (mohammadfawaz) [inline examples/debugging/src/lib.rs:53]: This space before the anchor is probably not needed?

File: docs/src/debugging/function-selector.md (hunk 4 above)
Comment (segfault-magnet) [reply docs/src/debugging/function-selector.md:1]: I've ran it through the default settings of [mdl](https://github.com/markdownlint/markdownlint). Suggested the empty lines around the code block.

Also complained about the line width. But I guess we would need to settle on a linter and the settings for it.

**Comment by segfault-magnet** [comment]
> This looks awesome, comments and tests are super on point! Left some nits/suggestions, plus one more thing: Can you write a very short section about this new feature in the book? This is very useful for SDK users diving into deeper levels (as opposed to writing tests, deploying, etc.), and it would be a PITA to learn about this functionality without an easy to find section in the book sweat_smile

I have no idea where to put it. For starters, I've added it to the debugging section.

**Comment by digorithm** [comment]
> > This looks awesome, comments and tests are super on point! Left some nits/suggestions, plus one more thing: Can you write a very short section about this new feature in the book? This is very useful for SDK users diving into deeper levels (as opposed to writing tests, deploying, etc.), and it would be a PITA to learn about this functionality without an easy to find section in the book sweat_smile
> 
> I have no idea where to put it. For starters, I've added it to the debugging section.

That's a good start; we can build on it later!

File: docs/src/debugging/function-selector.md (hunk 4 above)
Comment (mohammadfawaz) [reply docs/src/debugging/function-selector.md:1]: I usually use https://github.com/DavidAnson/markdownlint which has a VS code plugin. Wouldn't worry much about line width but vertical spaces are nice to have.

**Review by digorithm — APPROVED** [review]
LGTM!

**Pull request merged**